src/Core/Translation/DTO/MessageTranslation.php

```diff
@@ -0,0 +1,122 @@
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Core\Translation\DTO;
+
+/**
+ * This class is the representation of the last sheet of a translation catalogue : The message itself.
+ * A message is composed by the default wording,
+ * its translation within the project files (for crowdin or any translation tool),
+ * its translation made in the BO interface and stored in DB.
+ * If a message has file or user translation, it's considered as translated.
+ */
+class MessageTranslation
+{
+    /**
+     * @var string
+     */
+    private $defaultTranslation;
+
+    /**
+     * @var string|null
+     */
+    private $fileTranslation;
+
+    /**
+     * @var string|null
+     */
+    private $userTranslation;
+
+    public function __construct(string $defaultTranslation)
+    {
+        $this->defaultTranslation = $defaultTranslation;
+    }
+
+    public function getKey(): string
+    {
+        return $this->defaultTranslation;
+    }
+
+    public function setFileTranslation(string $fileTranslation): self
+    {
+        $this->fileTranslation = $fileTranslation;
+
+        return $this;
+    }
+
+    public function setUserTranslation(string $userTranslation): self
+    {
+        $this->userTranslation = $userTranslation;
+
+        return $this;
+    }
+
+    /**
+     * Returns whether a message is translated or not.
+     * It's TRUE if one of fileTranslation or userTranslation is not null
+     */
+    public function isTranslated(): bool
+    {
+        return null !== $this->fileTranslation || null !== $this->userTranslation;
+    }
+
+    /**
+     * Check if data contains search word.
+     *
+     * @param array $search
+     *
+     * @return bool
+     */
+    public function contains(array $search): bool
+    {
+        foreach ($search as $s) {
+            $s = strtolower($s);
+            if (
+                false !== strpos(strtolower($this->defaultTranslation), $s)
+                || (null !== $this->fileTranslation && false !== strpos(strtolower($this->fileTranslation), $s))
+                || (null !== $this->userTranslation && false !== strpos(strtolower($this->userTranslation), $s))
+            ) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /**
+     * @return array
+     */
+    public function toArray(): array
+    {
+        return [
+            'default' => $this->defaultTranslation,
+            'xliff' => $this->fileTranslation,
+            'database' => $this->userTranslation,
```

Comment:
```suggestion
            'user' => $this->userTranslation,
```

At first I thought user translation should be named databaseTranslation But then I thought it's actually better with user Because it reflects more the feature than the mean to attain it (it could be stored in file, API, whatever...) the important is that it's a custom translation made by the user (or merchant but user works better)

Which is why I think to be consistent we should use `user` everywhere instead of `database` so the array field here should be replaced Of course for implementation that clearly deal with database (like a DatabaseTranslationLoader, DoctrineProvider, or DbProvider, ...) the database word can be used, it's just one specific implementation of the User layer

Reply:
Since I don't know the details maybe you absolutely need this `database` key, in which case maybe you can just add a `user` key at least (which will be equal the the database one)

Reply:
Similarly `xliff` should be renamed into `file` But when I see this I'm actually doubting my own suggestion. Maybe the important here is not to know chat scope is targeted (default, file, user) but rather the technical mean to achieve it? (default, xliff, database, mongo, json, ...)

Reply:
Actually it depends on what we want to represent
file and database are storage system, xliff is a format, user is where the translation come from

I suggest to represent the source of the translation, no matter the way we store or extract it
- default
- project
- user

Reply:
Remains that ;)